File: Client/KataProfiler/core/Texture.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace KataProfiler
{
	public enum KPTextureType
	{
		TEX_NONE = 0,
		TEX_2D,
		TEX_CUBE_MAP
	}

	public class Texture : GlObject
	{
		public const int MAX_MIPMAP_LEVEL_NUMBER = 32;

		private Mipmap[] m_pMipmaps;
		public Mipmap[] Mipmaps { get { return m_pMipmaps; } }

		private KPTextureType m_texType;
		public KPTextureType TexType { get { return m_texType; } set { m_texType = value; } }

		public Texture() : base()
		{
			init();
		}

		public Texture(uint id) : base(id)
		{
			init();
		}

		private void init()
		{
			m_texType = KPTextureType.TEX_NONE;

			m_pMipmaps = new Mipmap[MAX_MIPMAP_LEVEL_NUMBER];
			for (int i = 0; i < MAX_MIPMAP_LEVEL_NUMBER; i++)
			{
				m_pMipmaps[i] = new Mipmap();
			}
		}

		public int MipmapCount
		{
			get
			{
				int count = 0;
				for (int i = 0; i < MAX_MIPMAP_LEVEL_NUMBER; i++)
				{
					if (m_pMipmaps[i].HasData) count++;
				}
				return count;
			}
		}

		public int SizeInBytes
		{
			get
			{
				int sum = 0;
				for (int i = 0; i < MAX_MIPMAP_LEVEL_NUMBER; i++) if (m_pMipmaps[i].HasData)
					{
						sum += m_pMipmaps[i].calculateSize();
					}
				return sum;
			}
		}

		public override void clearData()
		{
			m_texType = KPTextureType.TEX_NONE;
			for (int i = 0; i < MAX_MIPMAP_LEVEL_NUMBER; i++)
			{
				m_pMipmaps[i].reset();
			}
		}

		public override void fromMessage(Message msg)
		{
			clearData();

			MyBinStream stream = new MyBinStream(msg.Data);

			m_id = stream.readUInt();
			m_texType = (KPTextureType)stream.readInt();

			int mipmapCount = stream.readInt();

			for (int i = 0; i < mipmapCount; i++)
			{
				int level = stream.readInt();
				Utils.assert(level >= 0 && level < MAX_MIPMAP_LEVEL_NUMBER);
				Mipmap mip = m_pMipmaps[level];

				mip.reset();

				mip.Width = stream.readInt();
				mip.Height = stream.readInt();
				mip.IsCompressed = stream.readByte();
				mip.Format = stream.readInt();
				mip.Type = stream.readUInt();

				byte hasPixel = stream.readByte();
				if (hasPixel == 1)
				{
					int mipSize = mip.calculateSize();
					if (mipSize > 0)
					{
						mip.Pixels = new byte[mipSize];
						stream.readBytes(mip.Pixels, 0, mipSize);
					}
				}

				mip.HasData = true;
			}

			stream.close();
		}

		public void copyFrom(Texture tex)
		{
			clearData();
			m_id = tex.Id;
			m_texType = tex.TexType;

			Mipmap[] mips = tex.Mipmaps;
			for (int i = 0; i < MAX_MIPMAP_LEVEL_NUMBER; i++)
			{
				m_pMipmaps[i].copyFrom(mips[i]);
			}
		}

		public override bool isKindOf(KPGLObjectKind kind)
		{
			return kind == KPGLObjectKind.TEXTURE;
		}

		public void on_glCompressedTexImage2D(int level, uint internalformat, int width, int height,
					int border, int imageSize, byte[] data, int offset)
		{
			Mipmap mip = m_pMipmaps[level];
			mip.reset();

			mip.IsCompressed = 1;

			mip.Format = (int)internalformat;
			mip.Width = width;
			mip.Height = height;

			mip.Type = (uint)imageSize;

			if (imageSize > 0)
			{
				mip.Pixels = new byte[imageSize];
				Utils.memcpy(mip.Pixels, 0, data, offset, imageSize);
			}

			mip.HasData = true;
		}

		public void on_glTexImage2D(int level, int internalformat,
						int width, int height, int border, uint format, uint type, byte[] pixels, int offset)
		{
			Mipmap mip = m_pMipmaps[level];
			mip.reset();

			mip.IsCompressed = 0;

			mip.Format = (int)format;
			mip.Width = width;
			mip.Height = height;

			mip.Type = type;

			int imageSize = mip.calculateSize();
			if (imageSize > 0)
			{
				mip.Pixels = new byte[imageSize];
				Utils.memcpy(mip.Pixels, 0, pixels, offset, imageSize);
			}

			mip.HasData = true;
		}

		public void on_glTexStorage2D(int levels, int internalformat, int width, int height)
		{
			for (int level = 0; level < levels; ++level)
			{
				var mip = m_pMipmaps[level];
				mip.reset();

				mip.IsCompressed = Mipmap.isCompressedFormat(internalformat) ? (byte)1 : (byte)0;

				if (internalformat == gl.GL_RGB8)
					mip.Format = (int)gl.GL_RGB;
				else if (internalformat == gl.GL_RGBA8)
					mip.Format = (int)gl.GL_RGBA;
				else
					mip.Format = internalformat;

				mip.Type = mip.IsCompressed == 1 ? 16 : gl.GL_UNSIGNED_BYTE; // TODO

				mip.Width = width;
				mip.Height = height;

				int imageSize = mip.calculateSize();
				if (imageSize > 0)
				{
					mip.Pixels = new byte[imageSize];
					Utils.memset(mip.Pixels, 0, imageSize, 0);
				}

				mip.HasData = true;

				width = Math.Max(1, width / 2);
				height = Math.Max(1, height / 2);
			}
		}

		public void on_glTexSubImage2D(int level, int xoffset, int yoffset, int width, int height,
										uint format, uint type, byte[] pixels, int offset)
		{
			if (width <= 0 || height <= 0 || pixels.Length == offset) return;

			Mipmap mip = m_pMipmaps[level];

			if (mip.Width == 0 || mip.Height == 0 || mip.Pixels == null) return;

			int dst_bytesPP = mip.getBPP() / 8;
			int src_bytesPP = Mipmap.getTexBPP((int)format, type) / 8;

			for (int row = yoffset; row < yoffset + height; row++)
			{
				// TODO: row order???
				//int j2 = mip.Height - 1 - j;
				int row2 = row;
				for (int col = xoffset; col < xoffset + width; col++)
				{
					int pixelIndex = row2 * mip.Width + col;

					Utils.convertTexSubPixel(mip.Pixels, pixelIndex * dst_bytesPP, pixels, offset, format, mip.Type, type);

					//Utils.memcpy(mip.Pixels, pixelIndex * bytesPP, pixels, offset, bytesPP);

					offset += src_bytesPP;
				}
			}
		}

		public void on_glCompressedTexSubImage2D(int level, int xoffset, int yoffset,
													int width, int height, uint format, int imageSize, byte[] data, int offset)
		{
			// TODO
			Mipmap mip = m_pMipmaps[level];
			mip.Pixels = new byte[imageSize];
			Utils.memcpy(mip.Pixels, 0, data, offset, imageSize);
		}

	}
}

File: Server/src/Texture.cpp
#include "Texture.h"
#include "log.h"

Texture::Texture() : Texture(0)
{
}

Texture::Texture(u32 id) : GlObject(id), m_texType(TEX_NONE)
{
	for (int i = 0; i < MAX_MIPMAP; ++i)
	{
		m_mipmaps[i] = NULL;
	}
}

Texture::~Texture()
{
	clearData();
}

void Texture::clearData()
{
	for (int i = 0; i < MAX_MIPMAP; ++i)
	{
		SAFE_DEL(m_mipmaps[i]);
	}
}

TextureType Texture::getTexType()
{
	return m_texType;
}

void Texture::setTexType(TextureType texType)
{
	m_texType = texType;
}

int Texture::numMipmaps()
{
	int num = 0;
	for (int i = 0; i < MAX_MIPMAP; ++i)
	{
		if (m_mipmaps[i]) ++num;
	}
	return num;
}

Mipmap* Texture::on_glTexImage2D(GLenum target, GLint level, GLint internalformat,
	GLsizei width, GLsizei height, GLint border, GLenum format, GLenum type,
	const GLvoid* pixels, int unpackAlignment)
{
	if (!isValidTarget(target) || !isValidLevel(level))
		return NULL;
	SAFE_DEL(m_mipmaps[level]);
	m_mipmaps[level] = Mipmap::create_glTexImage2D(internalformat, width, height, border, format, type, pixels, unpackAlignment);
	return m_mipmaps[level];
}

Mipmap* Texture::on_glCompressedTexImage2D(GLenum target, GLint level, GLenum internalformat, GLsizei width, GLsizei height,
	GLint border, GLsizei imageSize, const GLvoid* data)
{
	if (!isValidTarget(target) || !isValidLevel(level))
		return NULL;
	SAFE_DEL(m_mipmaps[level]);
	m_mipmaps[level] = Mipmap::create_glCompressedTexImage2D(internalformat, width, height, border, imageSize, data);
	return m_mipmaps[level];
}

Mipmap* Texture::on_glCopyTexImage2D(GLenum target, GLint level, GLenum internalformat,
	GLsizei width, GLsizei height, GLint border, const char* rgba)
{
	if (!isValidTarget(target) || !isValidLevel(level))
		return NULL;
	SAFE_DEL(m_mipmaps[level]);
	m_mipmaps[level] = Mipmap::create_glCopyTexImage2D(internalformat, width, height, border, rgba);
	return m_mipmaps[level];
}

bool Texture::on_glTexStorage2D(GLenum target, GLsizei levels, GLenum internalformat, GLsizei width, GLsizei height)
{
	if (!isValidTarget(target))
		return false;
	if (levels <= 0 || !isValidLevel(levels - 1))
	{
		LOGE("[%s] Invalid levels %d", __FUNCTION__, levels);
		return false;
	}
	bool ok = true;
	for (int level = 0; level < levels; ++level)
	{
		SAFE_DEL(m_mipmaps[level]);
		m_mipmaps[level] = Mipmap::create_glTexStorage2D(internalformat, width, height);
		if (m_mipmaps[level] == NULL)
			ok = false;
		width = MAX(1, (width / 2));
		height = MAX(1, (height / 2));
	}
	return ok;
}

Mipmap* Texture::on_glTexSubImage2D(GLenum target, GLint level,
	GLint xoffset, GLint yoffset, GLsizei width, GLsizei height,
	GLenum format, GLenum type, const GLvoid* pixels, int unpackAlignment)
{
	if (!isValidTarget(target) || !isValidLevel(level) || !existsMipmap(level))
		return NULL;
	return m_mipmaps[level]->on_glTexSubImage2D(xoffset, yoffset, width, height, format, type, pixels, unpackAlignment);
}

Mipmap* Texture::on_glCompressedTexSubImage2D(GLenum target, GLint level, GLint xoffset, GLint yoffset,
	GLsizei width, GLsizei height, GLenum format, GLsizei imageSize, const GLvoid* data)
{
	if (!isValidTarget(target) || !isValidLevel(level))
		return NULL;
	
	if (m_mipmaps[level])
	{
		m_mipmaps[level]->on_glCompressedTexSubImage2D(xoffset, yoffset, width, height, format, imageSize, data);
		return m_mipmaps[level];
	}
	if (xoffset == 0 && yoffset == 0)
	{
		m_mipmaps[level] = Mipmap::create_glCompressedTexImage2D(format, width, height, 0, imageSize, data);
		return m_mipmaps[level];
	}
	LOGE("Mipmap with level %d does not exist", level);
	return NULL;
}

Mipmap* Texture::on_glCopyTexSubImage2D(GLenum target, GLint level, GLint xoffset, GLint yoffset,
	GLsizei width, GLsizei height, const char* rgba)
{
	if (!isValidTarget(target) || !isValidLevel(level) || !existsMipmap(level))
		return NULL;
	return m_mipmaps[level]->on_glCopyTexSubImage2D(xoffset, yoffset, width, height, rgba);
}

void Texture::copyRenderingRgba(int level, int width, int height, const char* pixels)
{
	if (!isValidLevel(level) || !existsMipmap(level)) return;
	m_mipmaps[level]->copyRenderingRgba(width, height, pixels);
}

Message* Texture::toMessage()
{
	Message* msg = new Message();
	msg->m_type = KMT_OBJECT_TEXTURE;

	int len = 4/*id*/ + 4/*texType*/ + 4/*mipmapCount*/;
	int mipmapCount = 0;
	for (int level = 0; level < MAX_MIPMAP; ++level)
	{
		if (m_mipmaps[level])
		{
			++mipmapCount;
			len += 4/*level*/ + 4/*width*/ + 4/*height*/ + 1/*isCompressed*/ + 4/*format*/ + 4/*type*/
				+ 1/*dummy*/ + m_mipmaps[level]->getImageSize();
		}
	}

	msg->m_length = len;
	msg->m_pData = new char[msg->m_length];
	char* buffer = msg->m_pData;
	memcpy(buffer, &m_id, 4);			buffer += 4;
	memcpy(buffer, &m_texType, 4);		buffer += 4;
	memcpy(buffer, &mipmapCount, 4);	buffer += 4;

	for (int level = 0; level < MAX_MIPMAP; ++level)
	{
		if (m_mipmaps[level])
		{
			Mipmap* mip = m_mipmaps[level];
			memcpy(buffer, &level, 4);			buffer += 4;
			int width = mip->getWidth();					memcpy(buffer, &width, 4);			buffer += 4;
			int height = mip->getHeight();					memcpy(buffer, &height, 4);			buffer += 4;
			u8 isCompressed = mip->isCompressed() ? 1 : 0;	memcpy(buffer, &isCompressed, 1);	buffer += 1;
			GLenum format = mip->getFormat();				memcpy(buffer, &format, 4);			buffer += 4;
			GLenum type = mip->getType();					memcpy(buffer, &type, 4);			buffer += 4;
			*buffer = 1;						buffer += 1;
			int imageSize = mip->getImageSize();
			memcpy(buffer, mip->getPixels(), imageSize);
			buffer += imageSize;
		}
	}
	return msg;
}

bool Texture::existsMipmap(int level)
{
	if (!m_mipmaps[level])
	{
		LOGE("Mipmap with level %d does not exist", level);
		return false;
	}
	return true;
}

bool Texture::isValidTarget(GLenum target)
{
	if (target != GL_TEXTURE_2D)
	{
		LOGE("Not supported target 0x%x, only GL_TEXTURE_2D (0x%x) is supported!", target, GL_TEXTURE_2D);
		return false;
	}
	return true;
}

bool Texture::isValidLevel(int level)
{
	if (level < 0 || level >= MAX_MIPMAP)
	{
		LOGE("Invalid level %d", level);
		return false;
	}
	return true;
}
